File: runtime/ruby/units/wasi/path_filestat_get.rb
# requires: memory/read_string, memory/init, wasi/resolve_path, wasi/pack_filestat, wasi/errno_fs
def wasi_path_filestat_get(dirfd, flags, path_ptr, path_len, buf_ptr)
  rel = @memory.read_string(path_ptr, path_len)
  host_path, err = resolve_path(dirfd, rel)
  return err if err
  symlink_follow = flags & 0x1 != 0 # lookupflags::SYMLINK_FOLLOW
  stat = symlink_follow ? File.stat(host_path) : File.lstat(host_path)
  @memory.init(buf_ptr, pack_filestat(stat), 0, 64)
  ERRNO_SUCCESS
rescue Errno::ENOENT
  ERRNO_NOENT
rescue SystemCallError
  ERRNO_IO
end

File: runtime/python/units/wasi/path_filestat_get.py
# requires: memory/read_string, memory/init, wasi/resolve_path, wasi/pack_filestat, wasi/errno_fs
def wasi_path_filestat_get(self, dirfd, flags, path_ptr, path_len, buf_ptr):
    rel = self.memory.read_string(path_ptr, path_len).decode("utf-8", "surrogateescape")
    symlink_follow = flags & 0x1 != 0  # lookupflags::SYMLINK_FOLLOW
    host_path, err = self.resolve_path(dirfd, rel, symlink_follow)
    if err is not None:
        return err
    try:
        st = os.stat(host_path) if symlink_follow else os.lstat(host_path)
    except OSError as e:
        return self.fs_errno(e)
    self.memory.init(buf_ptr, self.pack_filestat(st), 0, 64)
    return self.ERRNO_SUCCESS
